python/ray/train/_internal/session.py

```diff
@@ -269,22 +298,52 @@ def _report_thread_runner_error(self, block=False):
         except queue.Empty:
             pass
 
-    def checkpoint(self, **kwargs):
+    def checkpoint(self, checkpoint: Checkpoint):
         """Adds kwargs to the queue to be consumed by main thread.
 
         Also stores the checkpoint in ``self.loaded_checkpoint``.
         """
 
         # Update session checkpoint to latest checkpoint.
-        self.loaded_checkpoint = kwargs
+        self.loaded_checkpoint = checkpoint
+
+        encoded = False
 
         # Only store checkpoints on worker with rank 0.
         if self.world_rank != 0:
-            kwargs = {}
-        else:
-            kwargs = self._encode_data_fn(self._auto_fill_checkpoint_metrics(kwargs))
+            checkpoint = None
+        elif checkpoint and self._encode_data_fn:
+            checkpoint = checkpoint.from_dict(
+                self._encode_data_fn(checkpoint.to_dict())
+            )
+            encoded = True
+
+        if checkpoint and type(checkpoint) is Checkpoint:
+            checkpoint_dict = (
+                {}
+                if checkpoint.get_internal_representation()[0] != "data_dict"
+                else checkpoint.to_dict()
+            )
+            intended_checkpoint_class = self._get_checkpoint_class_fn(checkpoint_dict)
+            if intended_checkpoint_class is not Checkpoint:
+                if log_once("bad_checkpoint_type"):
+                    warnings.warn(
+                        f"You have reported a checkpoint with the `{Checkpoint}` "
+                        "type, but the intended checkpoint type for the Trainer "
+                        f"you are using is `{intended_checkpoint_class}`. Not using "
+                        "the intended checkpoint type may cause issues or "
+                        "exceptions, especially during serialization and "
+                        "deserialization. The checkpoint type will be changed "
+                        "automatically. This behavior may change in the future."
+                    )
+                checkpoint.__class__ = intended_checkpoint_class
```

Comment:
Hmm I'm not really following why we need this logic. Do we need this check at all?

How would this work for HorovodTrainer where both TorchCheckpoints and TensorflowCheckpoints are supported?

```
Not using "
"the intended checkpoint type may cause issues or "
"exceptions, especially during serialization and "
"deserialization. The checkpoint type will be changed "
"automatically.
```

what exactly is this referring to?

Reply:
This is for backwards compatibility reasons. If the class is not changed automatically, existing TorchTrainer/HorovodTrainer (with torch) code that uses base Checkpoint will break, as the encoding/decoding logic is now in TorchCheckpoint.

Only TorchCheckpoint has special encoding logic atm.